app/domains/__test__/TestCases.ts

```diff
@@ -1,14 +1,30 @@
-import type { Config } from "~/services/flow/server/types";
-import type { UserData } from "../userData";
+import { type Config } from "~/services/flow/server/types";
+import type { AllowedUserTypes, UserData } from "../userData";
 import { type ArrayConfigServer } from "~/services/array";
-import { type Guards } from "~/domains/guards.server";
 
 // Old flow tests: forward & backward using full user data
 export type TestCases<T extends UserData> = Readonly<
   Array<Readonly<[T, readonly string[]]>>
 >;
 
-export type ExpectedStep = {
+/**
+ * Retrieves template literal union of all possible array item property paths
+ * in the format arrayname#itemname
+ */
+type ArrayPropertyPaths<T> = {
+  [K in keyof T]: NonNullable<T[K]> extends ReadonlyArray<infer U>
+    ? U extends object
+      ? `${K & string}#${Extract<keyof U, string>}`
+      : never
+    : never;
+}[keyof T];
+
+// Deep partial needed to ensure partial array pages are handled correctly
+export type ExpectedStepUserInput<T extends UserData> = DeepPartial<T> & {
+  pageData?: { arrayIndexes?: number[] };
+} & { [K in ArrayPropertyPaths<T>]?: AllowedUserTypes };
```

Comment:
```suggestion
type ArrayItemPropertyTypes<T> = {
  [K in keyof T]: NonNullable<T[K]> extends ReadonlyArray<infer U>
    ? U extends object
      ? { [P in Extract<keyof U, string> as `${K & string}#${P}`]: U[P] }
      : {}
    : {};
}[keyof T];

export type ExpectedStepUserInput<T extends UserData> = T & {
  pageData?: { arrayIndexes?: number[] };
} & Partial<ArrayItemPropertyTypes<T>>;
```

I needed some help for this one, but it seems to work. Also `{}` merges more nicely than `never`.